include/bredis/AsyncConnection.hpp

```diff
@@ -0,0 +1,69 @@
+//
+//
+// Copyright (c) 2017 Ivan Baidakou (basiliscos) (the dot dmol at gmail dot com)
+//
+// Distributed under the MIT Software License
+//
+
+#pragma once
+
+#include <atomic>
+#include <initializer_list>
+#include <mutex>
+#include <queue>
+#include <sstream>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include <boost/asio.hpp>
+#include <boost/utility/string_ref.hpp>
+
+#include "Command.hpp"
+#include "Error.hpp"
+#include "Protocol.hpp"
+#include "Result.hpp"
+
+namespace bredis {
+
+template <typename AsyncStream> class AsyncConnection {
+    using protocol_type_t = typename AsyncStream::protocol_type;
+
+    static_assert(std::is_same<protocol_type_t, boost::asio::ip::tcp>::value ||
+                      std::is_same<protocol_type_t,
+                                   boost::asio::local::stream_protocol>::value,
+                  "AsyncConnection can be specialized either for ip::tcp or "
+                  "local::stream_protocol");
+
+  public:
+    using string_t = boost::string_ref;
+    using args_container_t = std::vector<string_t>;
+
+  private:
+    AsyncStream socket_;
```

Comment:
Consider renaming `socket_` to `stream_`. Rationale: `AsyncStream` (or `NextLayer`) might not be a socket. Users could be misled into thinking this only works with sockets.